query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PromQLMetricResponse.java

```diff
@@ -0,0 +1,93 @@
+package org.hypertrace.core.query.service.prometheus;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Singular;
+import lombok.Value;
+
+@Value
+class PromQLMetricResponse {
+  private static final ObjectMapper OBJECT_MAPPER =
+      new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+
+  @JsonProperty("status")
+  private String status = null;
+
+  @JsonProperty("data")
+  private PromQLData data = null;
+
+  @Getter
+  static class PromQLData {
+    @JsonProperty("resultType")
+    private String resultType;
+
+    @JsonProperty("result")
+    private List<PromQLMetricResult> result;
+  }
+
+  @Getter
+  @Builder
+  @AllArgsConstructor
+  @NoArgsConstructor
+  @EqualsAndHashCode
+  static class PromQLMetricResult {
```

Comment:
Here, Value + Builder with Singular gives me an error as Value makes it final and has to initialize. So, going with other existing notations.

Reply:
Hmm - not sure I follow. What's initializing? Is it a compile error? If the two can't be used together then make sense, but we should still use `@Value` by default in other classes (seen too many bugs on missing equals and hash codes!)

Reply:
OK pulled it down and looked. It's not the singular that's the issue, it's combining noargs constructor without a default value. So you've got a few options. You can use `@NoArgsConstructor(force = true)` to assign default values, or you can remove no arg constructor and use `@Jacksonized` to point jackson to the builder instead.

I'd expect you'd need something similar on the other classes too - PromQLMetricResponse and PromQLData. If I recall, by default Jackson looks for a no arg constructor which aren't provided on those classes.

Reply:
Thanks, incorporating the above change. I was not familiar with `jacksonized`.

Reply:
Done